pkgs/games/sm64ex/default.nix

```diff
@@ -1,71 +1,55 @@
-      "eu" = "c792e5ebcba34c8d98c0c44cf29747c8ee67e7b907fcc77887f9ff2523f80572";
-      "jp" = "9cf7a80db321b07a8d461fe536c02c87b7412433953891cdec9191bfad2db317";
-    }.${region};
-  }
+, callPackage
+, autoPatchelfHook
+, branch
 }:
 
-stdenv.mkDerivation rec {
-  pname = "sm64ex";
-  version = "unstable-2021-11-30";
-
-  src = fetchFromGitHub {
-    owner = "sm64pc";
-    repo = "sm64ex";
-    rev = "db9a6345baa5acb41f9d77c480510442cab26025";
-    sha256 = "sha256-q7JWDvNeNrDpcKVtIGqB1k7I0FveYwrfqu7ZZK7T8F8=";
+{
+  sm64ex = callPackage ./generic.nix {
+    pname = "sm64ex";
+    version = "0.pre+date=2021-11-30";
+
+    src = fetchFromGitHub {
+      owner = "sm64pc";
+      repo = "sm64ex";
+      rev = "db9a6345baa5acb41f9d77c480510442cab26025";
+      sha256 = "sha256-q7JWDvNeNrDpcKVtIGqB1k7I0FveYwrfqu7ZZK7T8F8=";
+    };
+
+    meta = {
+      homepage = "https://github.com/sm64pc/sm64ex";
+      description = "Super Mario 64 port based off of decompilation";
+    };
   };
 
-  nativeBuildInputs = [ python3 pkg-config ];
-  buildInputs = [ audiofile SDL2 hexdump ];
-
-  makeFlags = [ "VERSION=${region}" ] ++ compileFlags
-    ++ lib.optionals stdenv.isDarwin [ "OSX_BUILD=1" ];
-
-  inherit baseRom;
-
-  preBuild = ''
-    patchShebangs extract_assets.py
-    cp $baseRom ./baserom.${region}.z64
-  '';
-
-  installPhase = ''
-    mkdir -p $out/bin
-    cp build/${region}_pc/sm64.${region}.f3dex2e $out/bin/sm64ex
-  '';
-
-  enableParallelBuilding = true;
-
-  meta = with lib; {
-    homepage = "https://github.com/sm64pc/sm64ex";
-    description = "Super Mario 64 port based off of decompilation";
-    longDescription = ''
-      Super Mario 64 port based off of decompilation.
-      Note that you must supply a baserom yourself to extract assets from.
-      If you are not using an US baserom, you must overwrite the "region" attribute with either "eu" or "jp".
-      If you would like to use patches sm64ex distributes as makeflags, add them to the "compileFlags" attribute.
+  sm64ex-coop = callPackage ./generic.nix {
+    pname = "sm64ex-coop";
+    version = "0.pre+date=2022-05-14";
+
+    src = fetchFromGitHub {
+      owner = "djoslin0";
+      repo = "sm64ex-coop";
+      rev = "8200b175607fe2939f067d496627c202a15fe24c";
+      sha256 = "sha256-c1ZmMBtvYYcaJ/WxkZBVvNGVCeSXfm8NKe/BiAIJtks=";
+    };
+
+    nativeBuildInputs = [
+      autoPatchelfHook
+    ];
+
+    postInstall = let
+      sharedLib = stdenv.hostPlatform.extensions.sharedLibrary;
+    in ''
+      mkdir -p $out/lib
+      cp $src/lib/bass/libbass{,_fx}${sharedLib} $out/lib
+      cp $src/lib/discordsdk/libdiscord_game_sdk${sharedLib} $out/lib
     '';
-    license = licenses.unfree;
-    maintainers = with maintainers; [ ivar ];
-    platforms = platforms.unix;
+
+    meta = {
+      homepage = "https://github.com/djoslin0/sm64ex-coop";
+      description = "Super Mario 64 online co-op mod, forked from sm64ex";
+    };
   };
-}
+}.${branch}
```

Comment:
Can we avoid this `.${branch}` here? I think it is better to do something like `callPackage (builtins.getAttrs "sm64ex" (import ../games/sm64ex))` in `all-packages.nix` instead.

Reply:
Another option that may work too: `callPackage (inherit (import ../games/sm64ex) sm64ex) { }`.

Reply:
> Can we avoid this .${branch} here? I think it is better to do something like callPackage (builtins.getAttrs "sm64ex" (import ../games/sm64ex)) { } in all-packages.nix instead.

We need packages like `callPackage` and `autoPatchelfHook` in `sm64ex/default.nix`, not sure how to do that without it being a function called with `callPackage`. `builtins.getAttr` only seems to want to work if our import is an attrset instead of a function. Unless we inherit it in `all-packages.nix` but that doesn't sound very nice.

> Another option that may work too: callPackage (inherit (import ../games/sm64ex) sm64ex) { }.

Tried this as well, but hit a syntax error explaining `inherit` is unexpected here.

Reply:
Huh... Yeah, makes sense.

Maybe it would be better to split those two derivations them, having `sm64ex/default.nix` to be the original derivation and `sm64ex/coop.nix` to be the new one.

Reply:
I think that would be a bit annoying to maintain, seeing how there are many more forks of sm64ex which I intend to package in the future. Almost none of these have the building phase altered, so I think it'd be nice to inherit from a generic derivation like we do now so that we only need to swap out `src` and `name`.

Reply:
You could create a `sm64ex/common.nix` where you would export a function that contains everything shared between both of them, and them import this function in both `sm64ex/default.nix` and `sm64ex/coop.nix`.

I think this would arguably make it more easier to add future forks, instead of adding more and more in an increasingly large `sm64ex/default.nix` file.

Reply:
Actually, just saw that the way it currently is already looks kinda what I said (sorry, I forgot how this PR was before saying the previous message).

I still think maybe `default.nix` doesn't need to return an attrset and we can have `sm64ex/default.nix` pointing to the main repo and `sm64ex/coop.nix` being the coop fork, while both import the `sm64ex/generic.nix` to create their derivations. But I will approve as-is.